### ISSUE TITLE
Typescript: Actions in separate object resolve to wrong event type

### ISSUE THREAD

**Issue by DirkWolthuis** [davidkpiano/xstate#866]
**Description**
When setting actions in the state machine option object, the event type does not have the right property. It seems it uses the first event object from the Event type. 

**Expected Result**
I expected I can use any event object that I have defined in my Event object type. 

**Actual Result**
Typescript throws an error:
```
Property 'file' does not exist on type 'UploadFileEvent'.
  Property 'file' does not exist on type '{ type: "SUBMIT"; }'
```

**Reproduction**
```typescript
import {
    Machine,
    sendParent,
    State,
    SingleOrArray,
    OmniEvent,
    assign
} from "xstate";

export interface UploadFileStateSchema {
    states: {
        idle: {};
        incomplete: {};
    };
}

// The events that the machine handles
export type UploadFileEvent =
    | {
          type: "ADD_FILE";
          file: File;
      }
    | {
          type: "SUBMIT";
      };

export type UploadFileStateMachineContext = {
    file: File;
};

export type UploadFileStateMachineState = State<
    UploadFileStateMachineContext,
    UploadFileEvent
>;
export type UploadFileStateMachineSend = (
    event: SingleOrArray<OmniEvent<UploadFileEvent>>,
    payload?: Record<string, any> & {
        type?: undefined;
    }
) => State<UploadFileStateMachineContext, UploadFileEvent>;

export const uploadFileStateMachine = Machine<
    UploadFileStateMachineContext,
    UploadFileStateSchema,
    UploadFileEvent
>(
    {
        initial: "idle",
        states: {
            idle: {
                on: {
                    ADD_FILE: {
                        actions: "addFile"
                    },
                    SUBMIT: [
                        {
                            cond: "checkFile",
                            actions: "submit"
                        },
                        { target: "incomplete" }
                    ]
                }
            },
            incomplete: {
                on: {
                    ADD_FILE: {
                        actions: "addFile"
                    },
                    SUBMIT: [
                        {
                            cond: "checkFile",
                            actions: "submit"
                        },
                        { target: "incomplete" }
                    ]
                }
            }
        }
    },
    {
        actions: {
            submit: sendParent((context, event) => ({
                type: "SUBMIT_UPLOAD_FILE",
                file: context.file
            })),
            addFile: assign({
                file: (context, event) => event.file
            })
        },
        guards: {
            checkFile: (context, event) => !!context.file
        }
    }
);

```

**Additional context**
xstate 4.7.2, same error 4.6

**Comment by semopz** [comment]
see https://github.com/davidkpiano/xstate/issues/836#issuecomment-560154725

I know it's not ideal in your case since that action is referred to more than once. You could do this (which also isn't ideal)
```typescript
actions: {
            addFile: assign<UploadFileStateMachineContext , { type: "ADD_FILE", file: File }>({
                file: (context, event) => event.file
            })
        },
```

Not ideal obviously as you'd need to update your action if you later decide to make a change to your UploadFileEvent type

**Comment by DirkWolthuis** [comment]
Okay, I hope this will work someday. 

Here is my workaround for now: 

```
export type UploadFileEvent = AddFileEvent | SubmitEvent;

type AddFileEvent = {
    type: "ADD_FILE";
    file: File;
};
type SubmitEvent = {
    type: "SUBMIT";
};

addFile: assign({
    file: (context, event: AddFileEvent) => event.file
})
```

**Comment by Andarist** [comment]
At the moment TS inference fails us when it comes to inferring Context for `assign` and unfortunately once you specify a type argument you bail out from the inferring for all arguments imemdiately. So you cannot parametrize with Context, but leave Event to be inferred. There are issues about this on TS boards and they are figuring this out - but there is no ETA for this. So I'm afraid that at the moment there is no much we can do until that gets implemented in TS.

**Comment by DirkWolthuis** [comment]
> At the moment TS inference fails us when it comes to inferring Context for `assign` and unfortunately once you specify a type argument you bail out from the inferring for all arguments imemdiately. So you cannot parametrize with Context, but leave Event to be inferred. There are issues about this on TS boards and they are figuring this out - but there is no ETA for this. So I'm afraid that at the moment there is no much we can do until that gets implemented in TS.

Ok,  c'est la vie. Feel free to close this issue if you want to.

**Comment by roytouw7** [comment]
Any updates?

**Comment by xencodes** [comment]
no updates on this?

**Comment by davidkpiano** [comment]
> no updates on this?

You can use Typegen to resolve to the correct event types: https://stately.ai/blog/introducing-typescript-typegen-for-xstate

**Comment by weilliao-pelith** [comment]
In v5, we can use `setup()` to define action object. However, if the type definition of `event` has different payload, we will have to write more code to achieve type validation.

Here is an example. I fetched data outside the machine, and will tell actor to assign the fetched data to the context.
(just show the part of how to create the machine)
```ts
import { assign, setup } from 'xstate';

interface Context {
  chatRecord: { message: string } | null;
  loading: boolean;
  error: string;
}
 
// different events have different payload
type Event =
  | { type: 'FETCH_DATA' }
  | { type: 'DATA_RECEIVED'; data: { message: string } }
  | { type: 'FETCH_ERROR'; error: string };

const chatroomMachine = setup({
    types: {
        context: {} as Context
        events: {} as Event
    },
    actions: {
        onDataReceived: assign({
            // cause it's in action object, we have to check the event type and get the correct typing
            chatRecord: ({ context, event }) => {
                if (event.type === 'DATA_RECEIVED') {
                  return event.data;
                }
                return context.chatRecord
             },            
        })
    }
}).createMachine({
  id: 'chatroom-mahcine',
  initial: 'idle',
  context: {
    chatRecord: null,
    loading: false,
    error: '',
  },
  states: {
    idle: {
      on: {
        FETCH_DATA: 'loading',
      },
    },
    loading: {
        entry: assign({ loading: true }),
        on: {
            DATA_RECEIVED: {
                target: 'dataReceived',
                actions: 'onReceivedData'
            },
            FETCH_ERROR: {
                target: 'error',
                // inline action will get the payload automatically
                actions: assign({
                  error: ({ event }) => event.error,
                    loading: false
                })
            }
        }
    },
    dataReceived: {
        type: 'final'
    },
    error: {
        on: {
            FETCH_DATA: 'loading',
        }
    }
  },
})
```

**Comment by weilliao05621** [comment]
### Updates

TL;DR;
1. typegen is no longer needed in v5. check https://stately.ai/docs/setup#setting-up-types
2. if different event types have different payload, we need to handle type validation in action object in `setup`.


### Example
The following example describes more on 2.

In v5, we can use `setup()` to define action object. However, if the type definition of `event` has different payload, we will have to write more code to achieve type validation.

Here is an example. I fetched data outside the machine, and will tell actor to assign the fetched data to the context.
(just show the part of how to create the machine)
```ts
import { assign, setup } from 'xstate';

interface Context {
  chatRecord: { message: string } | null;
  loading: boolean;
  error: string;
}
 
// different events have different payload
type Event =
  | { type: 'FETCH_DATA' }
  | { type: 'DATA_RECEIVED'; data: { message: string } }
  | { type: 'FETCH_ERROR'; error: string };

const chatroomMachine = setup({
    types: {
        context: {} as Context
        events: {} as Event
    },
    actions: {
        onDataReceived: assign({
            // cause it's in action object, we have to check the event type and get the correct typing
            chatRecord: ({ context, event }) => {
                if (event.type === 'DATA_RECEIVED') {
                  return event.data;
                }
                return context.chatRecord
             },            
        })
    }
}).createMachine({
  id: 'chatroom-mahcine',
  initial: 'idle',
  context: {
    chatRecord: null,
    loading: false,
    error: '',
  },
  states: {
    idle: {
      on: {
        FETCH_DATA: 'loading',
      },
    },
    loading: {
        entry: assign({ loading: true }),
        on: {
            DATA_RECEIVED: {
                target: 'dataReceived',
                actions: 'onReceivedData'
            },
            FETCH_ERROR: {
                target: 'error',
                // inline action will get the payload automatically
                actions: assign({
                  error: ({ event }) => event.error,
                    loading: false
                })
            }
        }
    },
    dataReceived: {
        type: 'final'
    },
    error: {
        on: {
            FETCH_DATA: 'loading',
        }
    }
  },
})
```
